src/BaseForm.php

```diff
@@ -241,4 +248,35 @@ private function invokeHandler($handler, $formValues)
 
 		$handler(...$params);
 	}
+	
+	public function disallow4ByteCharacters(string $errorMessage)
+	{
+		$this->disallow4ByteCharacterMessage = $errorMessage;
+	}
+
+	/**
+	 * don't allow UTF8 4 bytes characters in TextBase controls
+	 * if BaseForm::disallow4ByteCharacterMessage is set
+	 * and TextBase::setOption(self::OPTION_ALLOW_4_BYTE_CHARACTERS, true) is not set
+	 */
+	private function add4ByteCharacterRule(Form $form)
+	{
+		if (!$this->disallow4ByteCharacterMessage) {
+			return;
+		}
+
+		foreach ($form->getControls() as $_control) {
+			if (!$_control instanceof TextBase) {
+				continue;
+			}
+
+			if ($_control->getOption(self::OPTION_ALLOW_4_BYTE_CHARACTERS, false)) {
+				continue;
+			}
+
+			$_control->addRule(function(TextBase $control) {
+				return !preg_match_all('~[\x{10000}-\x{10FFFF}]~u', $control->getValue());
```

Comment:
Tady by mohlo stačit pouze `preg_match`?

Reply:
logicky by melo no...